ballerina/websocket_utils.bal

```diff
@@ -0,0 +1,123 @@
+            check sendWebSocketResponse(caller, customHeaders, WS_ERROR, errorPayload, connectionId);
+            closeConnection(caller);
+        } else {
+            check sendWebSocketResponse(caller, customHeaders, WS_COMPLETE, null, connectionId);
+        }
+    } else {
+        check sendWebSocketResponse(caller, customHeaders, WS_ERROR, sourceStream, connectionId);
+        closeConnection(caller);
+    }
+}
+
+isolated function validateSubscriptionPayload(json|WSPayload data, Engine engine) returns parser:OperationNode|json {
+    json|error payload = data is WSPayload ? data?.payload : data;
+    if payload is error {
+        return {errors: [{message: "Invalid format in WebSocket payload: " + payload.message()}]};
+    }
+    json|error document = payload.query;
+    if document is error {
+        return {errors: [{message: "Unable to find the query: " + document.message()}]};
+    }
+    if document !is string {
+        return {errors: [{message: "Invalid format in request parameter: query"}]};
+    }
+    if document == "" {
+        return {errors: [{message: "Empty query is found"}]};
+    }
+    json|error variables = payload.variables;
+    variables = variables is error && variables.message() == "{ballerina/lang.map}KeyNotFound" ? () : variables;
+    if variables is error || variables !is map<json>? {
+        return {errors: [{message: "Invalid format in request parameter: variables"}]};
+    }
+    parser:OperationNode|OutputObject validation = engine.validate(document, getOperationName(payload), variables);
+    if validation is parser:OperationNode {
+        return validation;
+    }
+    return validation.toJson();
+}
+
+isolated function getSubscriptionResponse(Engine engine, __Schema schema, Context context,
+                                          parser:FieldNode node) returns stream<any, error?>|json {
+    ExecutorVisitor executor = new(engine, schema, context, {});
+    any|error result = getSubscriptionResult(executor, node);
+    if result !is stream<any, error?> {
+        string errorMessage = result is error ? result.message() : "Invalid return type. Expected type: stream";
+        return {errors: [{message: errorMessage}]};
+    }
+    return <stream<any, error?>>result;
+}
+
+isolated function sendWebSocketResponse(websocket:Caller caller, map<string> & readonly customHeaders, string wsType,
+                                        json payload, string? id = ()) returns websocket:Error? {
+    if !customHeaders.hasKey(WS_SUB_PROTOCOL) {
+        check caller->writeMessage(payload);
+    } else {
+        string 'type = getMessageType(wsType, customHeaders);
+        json jsonResponse = id != () ? {'type: 'type, id: id, payload: payload} : {'type: 'type, payload: payload};
+        check caller->writeMessage(jsonResponse);
+    }
+}
+
+isolated function getMessageType(string wsType, map<string> & readonly customHeaders) returns string {
+    if customHeaders.get(WS_SUB_PROTOCOL) == GRAPHQL_WS {
```

Comment:
This will panic if the `WS_SUB_PROTOCOL` is not found in the map. Should first use the `hasKey` function to check whether the key is present.

Reply:
It is already checked before the `getMessageType` method is called.